# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

### My Prediction

**I predict Random Forest will perform better than Logistic Regression.  In general, Random Forest seems more robust than Logistic Regression in modeling connections between the predictor values and the result.  Random Forest uses predictions of numerous decision trees to make a final prediction, it handles linear and non-linear relationships between the variables, and handles high-dimensional data with less risk of overfitting.  Also, in reviewing the dataset information, the Baseline Model Performance illustrates that the Random Forest model is more accurate and precise than the Logistic Regression model for this dataset.**

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`

# Create the labels set `y`
y = data['spam']
display(y[0:5])

# Create features DataFrame `X`
X = data.copy()
X = X.drop(columns="spam")
display(X.head())

0    1
1    1
2    1
3    1
4    1
Name: spam, dtype: int64

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.00  ...                   0.0         0.00   
1             0.00            0.94  ...                   0.0         0.00   
2             0.64            0.25  ...                   0.0         0.01   
3             0.31            0.63  ...                   0.0         0.00   
4             0.31            0.63  ...                   0.0         0.00   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0        0.000          0.0        0.778        0.000        0.000   
1        0.132          0.0        0.372        0.180        0.048   
2        0.143          0.0        0.276        0.184        0.010   
3        0.137          0.0        0.137        0.000        0.000   
4        0.135          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  
0                       278  
1                      1028  
2                      2259  
3                       191  
4                       191  

[5 rows x 57 columns]

In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [7]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)


StandardScaler()

In [8]:
# Scale the training and test data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [9]:
# Display the scaled training data
X_train_scaled[:5]

array([[-0.33933373, -0.16214839, -0.55144324, -0.0455687 , -0.45260014,
        -0.3517885 ,  3.0756918 , -0.25647335, -0.31841516, -0.37271758,
        -0.29501265, -0.62180957, -0.32657781, -0.17617526, -0.18908381,
         1.09187591, -0.31666446, -0.34504726, -0.93039862, -0.154063  ,
        -0.66157193, -0.11522182, -0.28644389, -0.20366572, -0.32352185,
        -0.30020799, -0.22881004, -0.23108839, -0.17233002, -0.22127159,
        -0.17260004, -0.1416128 , -0.1921702 , -0.14436226, -0.19157879,
        -0.24075252, -0.33092557, -0.05564806, -0.17725653, -0.18198114,
        -0.12101748, -0.17228913, -0.20224232, -0.12847784, -0.29799633,
        -0.2104819 , -0.07057583, -0.11540959, -0.15779434, -0.48712282,
        -0.17291703,  1.90383911, -0.31944577, -0.12204215, -0.11649103,
        -0.21175305, -0.41909144],
       [ 1.8777157 , -0.16214839, -0.55144324, -0.0455687 , -0.45260014,
        -0.3517885 , -0.29726709, -0.25647335, -0.31841516, -0.37271758,
        -0.29501

In [10]:
# Display the scaled testing data
X_test_scaled[:5]

array([[-0.33933373, -0.16214839, -0.55144324, -0.0455687 , -0.45260014,
        -0.3517885 , -0.29726709, -0.25647335, -0.31841516, -0.37271758,
        -0.29501265, -0.62180957, -0.32657781, -0.17617526, -0.18908381,
        -0.29202679, -0.31666446, -0.34504726, -0.93039862, -0.154063  ,
        -0.66157193, -0.11522182, -0.28644389, -0.20366572, -0.32352185,
        -0.30020799, -0.22881004, -0.23108839, -0.17233002, -0.22127159,
        -0.17260004, -0.1416128 , -0.1921702 , -0.14436226, -0.19157879,
        -0.24075252, -0.33092557, -0.05564806, -0.17725653, -0.18198114,
        -0.12101748, -0.17228913, -0.20224232, -0.12847784, -0.29799633,
        -0.2104819 , -0.07057583, -0.11540959, -0.15779434, -0.48712282,
        -0.17291703, -0.30668088, -0.31944577, -0.12204215, -0.12843428,
        -0.2444289 , -0.48382584],
       [ 2.01007686, -0.16214839,  0.84057643, -0.0455687 , -0.45260014,
        -0.3517885 , -0.29726709, -0.25647335, -0.31841516,  0.74260757,
        -0.29501

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [11]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

# Create a `LogisticRegression` function
lr_model = LogisticRegression(random_state=1)

# Fit the model
lr_model.fit(X_train_scaled, y_train)

# Print the model score
print(f"Training Data Score: {lr_model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {lr_model.score(X_test_scaled, y_test)}")

Training Data Score: 0.9257971014492754
Testing Data Score: 0.9226759339704604


In [12]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = lr_model.predict(X_test_scaled)

# Review the predictions
testing_predictions

array([0, 0, 0, ..., 1, 0, 0], dtype=int64)

In [13]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
lr_acc_score = accuracy_score(y_test, testing_predictions)

# Display results
print(f"Logistic Regression Accuracy Score : {lr_acc_score}")

Logistic Regression Accuracy Score : 0.9226759339704604


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [14]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

# Create the random forest classifier instance
rf_model = RandomForestClassifier(n_estimators=128, random_state=1)

# Fit the model
rf_model.fit(X_train_scaled, y_train)

# Print the model score
print(f"Training Data Score: {rf_model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {rf_model.score(X_test_scaled, y_test)}")


Training Data Score: 0.9994202898550725
Testing Data Score: 0.9609035621198957


In [15]:
# Make and save testing predictions with the saved logistic regression model using the test data

# Make predictions using the testing data
rf_testing_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
rf_testing_predictions

array([0, 0, 0, ..., 1, 0, 0], dtype=int64)

In [16]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_acc_score = accuracy_score(y_test, rf_testing_predictions)

# Display Random Forest Accuracy
print(f"Random Forest Accuracy Score : {rf_acc_score}")

Random Forest Accuracy Score : 0.9609035621198957


In [17]:
# Display Logistic Regression Accuracy again for comparison
print(f"Logistic Regression Accuracy Score : {lr_acc_score}")

Logistic Regression Accuracy Score : 0.9226759339704604


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

### My Evaluation

***Question 1:  Which model performed better?*** 

**Answer:  The Random Forest model performed better on this data than Logistic Regression.  The Random Forest Accuracy Score is 0.961 and Logistic Regression Accuracy Score is 0.923.**
 
***Question 2:  How does that compare to your prediction?*** 

**Answer:  I predicted Random Forest would perform better and the results match my prediction.**